### ISSUE TITLE
Released packages should have corresponding Git tags

### ISSUE THREAD

**Issue by wincent** [liferay/liferay-npm-tools#40]
There are no tags in this repo, but we should be making and pushing a tag every time we make a release. If I go to the official package pages, I should be able to click the GitHub link and in this repo find the code that corresponds to the package (via the tag):

- https://www.npmjs.com/package/liferay-npm-bundler-preset-liferay-dev
- https://www.npmjs.com/package/liferay-jest-junit-reporter
- https://www.npmjs.com/package/liferay-npm-scripts

Some of these we can go back and add retroactively, based on the commit messages, but others may be harder to figure out. I'd say that the most important thing would be to just make sure we tag new releases going forward; dealing with the historical versions isn't super critical.

One thing to bear in mind: because we release the packages independently, the tags need to be "namespaced" by package name. That is, we should make these tags:

- `liferay-npm-scripts/v1.4.5` pointing at https://github.com/liferay/liferay-npm-tools/commit/a90f91b0a410df87a3d66eec836fe817ca35321b
- `liferay-npm-scripts/v1.4.4` pointing at https://github.com/liferay/liferay-npm-tools/commit/bbc2a8afddfcd879ddbb0de314c1bf21a896b1ec
- `liferay-npm-scripts/v1.4.3` pointing at https://github.com/liferay/liferay-npm-tools/commit/5bc85812f7377d3758226035e0d247d2ae63f23b
- `liferay-npm-bundler-preset-liferay-dev/v1.1.3` pointing at https://github.com/liferay/liferay-npm-tools/commit/c82462cdc7805c840e1b1f1b57f510d21b536adb
- `liferay-npm-scripts/v1.4.2` pointing at https://github.com/liferay/liferay-npm-tools/commit/f0e1b037cdff5676d93d9f771b8b4a9bf461925f
- `liferay-npm-scripts/v1.4.1` pointing at https://github.com/liferay/liferay-npm-tools/commit/0c78b0af8fbb4b3786b68c493ae4d4fb4242ce6c
- `liferay-npm-scripts/v1.4.0` pointing at https://github.com/liferay/liferay-npm-tools/commit/2edbc6ff176f6f16807bf1d89365ba6cf24ace42 (ironically, that commit is titled "Prepare liferay-npm-scripts v1.3.0").
- `liferay-npm-scripts/v1.3.0` pointing at https://github.com/liferay/liferay-npm-tools/commit/94930736ab79f41ae487736c66432499b88c0752
- `liferay-npm-scripts/v1.2.0` pointing at https://github.com/liferay/liferay-npm-tools/commit/8ca4c78cc5a3025a6c58fdeaff8b03b22b234878
- `liferay-npm-scripts/v1.1.0` pointing at https://github.com/liferay/liferay-npm-tools/commit/a3f7ef2dcd7015b65fbecebfd62d8d74e0f8146a
- `liferay-jest-junit-reporter/v1.0.0` pointing at https://github.com/liferay/liferay-npm-tools/commit/e906a4ca14db441727bec323ce261153d05f66ed
- `liferay-npm-scripts/v1.0.0` pointing at https://github.com/liferay/liferay-npm-tools/commit/878f1be17897b9d2058b81ba5c50cf318d645aff
- `liferay-npm-scripts/v0.0.8` pointing at https://github.com/liferay/liferay-npm-tools/commit/c0a68c899a2dc86be09d974ff0852e2ce81f53e1
- `liferay-npm-scripts/v0.0.7` pointing at https://github.com/liferay/liferay-npm-tools/commit/300040e589cca3be040046a40136aac578797dee
- `liferay-npm-scripts/v0.0.6` pointing at https://github.com/liferay/liferay-npm-tools/commit/a7c158774325eb038f0cf2ffa50a249ccada8af0
- `liferay-npm-scripts/v0.0.5` pointing at https://github.com/liferay/liferay-npm-tools/commit/91a4353a1b655400e562a81d19438d22abf0eecd
- `liferay-npm-scripts/v0.0.4` pointing at https://github.com/liferay/liferay-npm-tools/commit/40fb324ce4a4b8e444bb05ba706ddba3e57d7641
- `liferay-npm-scripts/v0.0.3` pointing at https://github.com/liferay/liferay-npm-tools/commit/2d14dfa306d777d09fef852bf6a80565cafd90b9
- `liferay-npm-scripts/v0.0.2` pointing at https://github.com/liferay/liferay-npm-tools/commit/13fb266981a058f7892e94837154f5d23a55426a
- `liferay-jest-junit-reporter/v0.0.4` pointing at https://github.com/liferay/liferay-npm-tools/commit/0b30700b6777e8374e7bb643b01c3ea6c0f3eb38
- `liferay-jest-junit-reporter/v0.0.3` pointing at https://github.com/liferay/liferay-npm-tools/commit/0ab6999307ea782e7ddbc973c1efec94f9d613af
- `liferay-jest-junit-reporter/v0.0.2` pointing at https://github.com/liferay/liferay-npm-tools/commit/ae47b29e0432403df1f8a07eb0074d7f06004554

**Comment by wincent** [comment]
Created and pushed the tags using:

```
git tag -m 'liferay-npm-scripts v1.4.5' liferay-npm-scripts/v1.4.5 a90f91b
git tag -m 'liferay-npm-scripts v1.4.4' liferay-npm-scripts/v1.4.4 bbc2a8a
git tag -m 'liferay-npm-scripts v1.4.3' liferay-npm-scripts/v1.4.3 5bc8581
git tag -m 'liferay-npm-bundler-preset-liferay-dev v1.1.3' liferay-npm-bundler-preset-liferay-dev/v1.1.3 c82462c
git tag -m 'liferay-npm-scripts v1.4.2' liferay-npm-scripts/v1.4.2 f0e1b03
git tag -m 'liferay-npm-scripts v1.4.1' liferay-npm-scripts/v1.4.1 0c78b0a
git tag -m 'liferay-npm-scripts v1.4.0' liferay-npm-scripts/v1.4.0 2edbc6f
git tag -m 'liferay-npm-scripts v1.3.0' liferay-npm-scripts/v1.3.0 9493073
git tag -m 'liferay-npm-scripts v1.2.0' liferay-npm-scripts/v1.2.0 8ca4c78
git tag -m 'liferay-npm-scripts v1.1.0' liferay-npm-scripts/v1.1.0 a3f7ef2
git tag -m 'liferay-jest-junit-reporter v1.0.0' liferay-jest-junit-reporter/v1.0.0 e906a4c
git tag -m 'liferay-npm-scripts v1.0.0' liferay-npm-scripts/v1.0.0 878f1be
git tag -m 'liferay-npm-scripts v0.0.8' liferay-npm-scripts/v0.0.8 c0a68c8
git tag -m 'liferay-npm-scripts v0.0.7' liferay-npm-scripts/v0.0.7 300040e
git tag -m 'liferay-npm-scripts v0.0.6' liferay-npm-scripts/v0.0.6 a7c1587
git tag -m 'liferay-npm-scripts v0.0.5' liferay-npm-scripts/v0.0.5 91a4353
git tag -m 'liferay-npm-scripts v0.0.4' liferay-npm-scripts/v0.0.4 40fb324
git tag -m 'liferay-npm-scripts v0.0.3' liferay-npm-scripts/v0.0.3 2d14dfa
git tag -m 'liferay-npm-scripts v0.0.2' liferay-npm-scripts/v0.0.2 13fb266
git tag -m 'liferay-jest-junit-reporter v0.0.4' liferay-jest-junit-reporter/v0.0.4 0b30700
git tag -m 'liferay-jest-junit-reporter v0.0.2' liferay-jest-junit-reporter/v0.0.3 0ab6999
git tag -m 'liferay-jest-junit-reporter v0.0.2' liferay-jest-junit-reporter/v0.0.2 ae47b29
git push upstream-rw --tags
```

**Comment by wincent** [comment]
Mentioning you here @jbalsas and @bryceosterhaus  because I think you've both cut a few releases. Should I put a reminder about tagging releases in a `CONTRIBUTING.md` at the top of the repo?

**Comment by jbalsas** [comment]
❤️ Thanks @wincent! Yeah, I noticed yesterday we weren't tagging, and I'm also not sure we have a common merge strategy here either...

**Comment by bryceosterhaus** [comment]
Good catch! Didn't realize it.

As for merge strategy, is there any other guidelines we follow for other projects? It'd be nice to outline general contributing rules for portal and non-portal related gh projects

**Comment by wincent** [comment]
@jbalsas: When you say merge strategy, are you referring to how/when/if to merge develop and master (and in which direction)?

I know that in most but not all Liferay projects outside of portal we follow — more or less — the "Gitflow" workflow described here:

- https://nvie.com/posts/a-successful-git-branching-model/
- https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow

Examples:

- [Clay](https://github.com/liferay/clay) uses it, as does [liferay-npm-build-tools](https://github.com/liferay/liferay-npm-build-tools) and I suspect many others.
- Some projects are very simple and work just from "master", like [eslint-config-liferay](https://github.com/liferay/eslint-config-liferay).
- [liferay-amd-loader](https://github.com/liferay/liferay-amd-loader) is an example of a project which doesn't use "Gitflow" (just works off master), but also has release branches (eg. [2.x](https://github.com/liferay/liferay-amd-loader/tree/2.x) and [3.x](https://github.com/liferay/liferay-amd-loader/tree/3.x), which TBH look seldom updated).

I am pretty skeptical of "Gitflow" myself. I think for simple projects it brings unnecessary complexity, and for more complicated projects, er, it probably bring unnecessary complexity there too. It's all too easy to send a pull request to the wrong branch, and then we have confusion like we've seen in Clay where people [apparently don't know where the work is happening](https://github.com/liferay/clay/pull/1674).

In both [alloy-editor](https://github.com/liferay/alloy-editor) and [liferay-js-themes-toolkit](https://github.com/liferay/liferay-js-themes-toolkit) we actually moved away from "Gitflow" and back to a more conventional develop-on-master approach. We still have a branch that we update after ever release, but it's just not called "master" any more:

- In AlloyEditor:
  - new development work for v2.0 happens in "master".
  - after each release "stable" gets fast-forwarded to match "master".
  - work on the legacy v1.x versions happens on "1.x" (we still need to cut releases there occasionally).
  - after each v1.x release "1.x-stable" gets fast-forwarded to match "1.x".
- In the themes toolkit:
  - new development work for v9.0 happens in "master".
  - after each release "stable" gets fast-forwarded to match "master".
  - work on the legacy v8.x versions happens on "8.x" (we still need to cut releases there occasionally).
  - after each v8.x release "8.x-stable" gets fast-forwarded to match "8.x".

Basically, we're optimizing for making the common case as easy as possible: that is, the bulk of our work happens on "master". I think that these projects where we need to maintain multiple long-lived release branches are more likely the exception than the rule, and most of our projects could just use a simple "do it on master" model. Even the value of the "stable" branch is debatable, although it doesn't cost that much to keep it around.

(And now that my portal build has finally finished, I will go away and stop commenting on GitHub.)

**Comment by wincent** [comment]
@julien: I don't know if you have some thoughts on branching, as you've been involved in a couple of those projects where we changed the branching model.

**Comment by julien** [comment]
@wincent I like what we've been using in **alloy-editor** and **liferay-js-themes-toolkit** at least in those project where we had big changes, it was useful and easy enough for me to remember.

**Comment by wincent** [comment]
Created https://github.com/liferay/liferay-npm-tools/issues/47 so we have a place to discuss merge strategies, and closing this one for now because there's nothing actionable remaining.
